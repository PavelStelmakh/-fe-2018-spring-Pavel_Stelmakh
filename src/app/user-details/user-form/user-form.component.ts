import { Component, OnInit, Output, EventEmitter, Input } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { usernameValidator, userageValidator, dateValidator, checkExistNameValidator } from '../../validators';
import { UsersService } from '../../users.service';
import { User } from '../../../../models/User';
import { PopupService } from '../../popup/popup.service';
import * as moment  from 'moment';
import { Store, select } from '@ngrx/store';
import * as usersAction from '../../actions/users.action';
import * as usersReducer from '../../reducers/users.reducer';
import * as profileAction from '../../actions/profile.action';
import * as profileReducer from '../../reducers/profile.reduser';

@Component({
  selector: 'app-user-form',
  templateUrl: './user-form.component.html',
  styleUrls: ['./user-form.component.scss']
})
export class UserFormComponent implements OnInit {
  userForm: FormGroup;
  private id: number;
  @Output() update: EventEmitter<number> = new EventEmitter<number>();
  @Input('edit') isEditUser: boolean;
  private isCreateNewUser: boolean;

  constructor(
    private builder: FormBuilder, 
    private user: UsersService,
    private popup: PopupService,
    private storeUsers: Store<usersReducer.State>,
    private storeProfile: Store<profileReducer.State>
    ) {}

  ngOnInit() {
    this.isCreateNewUser = false;
    this.userForm = this.builder.group({
      name: ['', Validators.required],
      password: ['', [Validators.required, Validators.minLength(3)]],
      age: ['', [Validators.required, userageValidator()]],
      dateOfBirth: ['', [Validators.required, dateValidator('YYYY/MM/DD')]],
      dateOfFirstLogin: ['', [Validators.required, dateValidator('DD MMMM YYYY')]],
      information: ['', [Validators.required]],
      dateOfNextNotif: ['', [Validators.required, dateValidator('DD-MMM-YY')]]
    });
    if (!this.isEditUser) {
      this.storeProfile.pipe(select(profileReducer.getProfileUser)).subscribe((value: User) => this.setFormValue(value));
    } else {
      this.storeUsers.pipe(select(usersReducer.getSelectedUser)).subscribe((value: User) => this.setFormValue(value));
    }
  }

  get name() {
    return this.userForm.get('name');
  }

  get password() {
    return this.userForm.get('password');
  }

  get age() {
    return this.userForm.get('age');
  }

  get dateOfBirth() {
    return this.userForm.get('dateOfBirth');
  }

  get dateOfNextNotif() {
    return this.userForm.get('dateOfNextNotif');
  }

  get information() {
    return this.userForm.get('information');
  }

  get dateOfFirstLogin() {
    return this.userForm.get('dateOfFirstLogin');
  }

  onSubmit() {
    const value: User = this.userForm.value;
    if (!this.isCreateNewUser) {
      value.id = this.id;
      this.popup.show();
      if (!this.isEditUser) {
        this.update.emit(0);
        this.storeProfile.dispatch(new profileAction.EditProfileAction(value));
      } else {
        this.update.emit(1);
        this.storeUsers.dispatch(new usersAction.EditUserAction(value));
      } 
    } else {
      this.storeUsers.dispatch(new usersAction.AddUserAction(value));
      this.popup.show();
      this.update.emit(1); 
    }
  }

  setFormValue(value: User) {
    if (value) {
      this.isCreateNewUser = false;
      this.id = value.id;
      this.name.setAsyncValidators([usernameValidator(), checkExistNameValidator(this.user, false, this.id)]);
      const dateOfNextNotif = moment(value.dateOfNextNotif).format('DD-MMM-YY');
      this.userForm.patchValue({
        name: value.name,
        password: value.password,
        age: value.age,
        dateOfBirth: moment(value.dateOfBirth).format('YYYY/MM/DD'),
        dateOfFirstLogin: moment(value.dateOfFirstLogin).format('DD MMMM YYYY'),
        information: value.information,
        dateOfNextNotif: dateOfNextNotif !== 'Invalid date' ? dateOfNextNotif: value.dateOfNextNotif
      });
    } else {
      this.isCreateNewUser = true;
      this.name.setAsyncValidators([usernameValidator(), checkExistNameValidator(this.user, false)]);
    }
  }

}
