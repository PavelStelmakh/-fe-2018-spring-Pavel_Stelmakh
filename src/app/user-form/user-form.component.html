<form [formGroup]="userForm" (ngSubmit)="onSubmit()">
    <div>
        <label>
            <span>{{'user-details.user-form.name' | translate}}</span>
            <input type="text" placeholder="Anna Maria" formControlName="name">
        </label>
        <div class="loading" *ngIf="name.dirty && name.status === 'PENDING'">loading...</div>
        <app-invalid-input *ngIf="name.dirty && name.status === 'VALID'" 
        message="{{'user-details.user-form.messages.success' | translate}}"
        type="success"></app-invalid-input>
        <app-invalid-input *ngIf="name.dirty && (name.errors?.name || name.errors?.checkExistName)" 
        message="{{ name.errors?.name ? ('user-details.user-form.messages.name' | translate) 
        : ('user-details.user-form.messages.existname' | translate) }}"
        type="error"></app-invalid-input>
    </div>
    <div>
        <label>
            <span>{{'user-details.user-form.age' | translate}}</span>
            <input type="text" formControlName="age" placeholder="18-65">
        </label>
        <app-invalid-input *ngIf="age.dirty && (age.errors?.age || age.status === 'VALID')" 
        message="{{ age.status === 'VALID' ? ('user-details.user-form.messages.success' | translate)
        : ('user-details.user-form.messages.age' | translate) }}"
        type="{{ age.status === 'VALID' ? 'success' : 'error'}}"></app-invalid-input>
    </div>
    <div>
        <label>
            <span>{{'user-details.user-form.birthday' | translate}}</span>
            <input type="text" formControlName="dateOfBirth" placeholder="YYYY/MM/DD">
        </label>
        <app-invalid-input *ngIf="dateOfBirth.dirty && (dateOfBirth.errors?.date || dateOfBirth.status === 'VALID')" 
        message="{{ dateOfBirth.status === 'VALID' ? ('user-details.user-form.messages.success' | translate) 
        : ('user-details.user-form.messages.dateOfBirth' | translate) }}"
        type="{{ dateOfBirth.status === 'VALID' ? 'success' : 'error'}}"></app-invalid-input>
    </div>
    <div>
        <label>
            <span>{{'user-details.user-form.dateOfLogin' | translate}}</span>
            <input type="text" formControlName="dateOfFirstLogin" placeholder="DD MMMM YYYY">
        </label>
        <app-invalid-input 
        *ngIf="dateOfFirstLogin.dirty && (dateOfFirstLogin.errors?.date || dateOfFirstLogin.status === 'VALID')" 
        message="{{ dateOfFirstLogin.status === 'VALID' ? ('user-details.user-form.messages.success' | translate) 
        : ('user-details.user-form.messages.dateOfFirstLogin' | translate) }}"
        type="{{ dateOfFirstLogin.status === 'VALID' ? 'success' : 'error'}}"></app-invalid-input>
    </div>
    <div>
        <label>
            <span>{{'user-details.user-form.information' | translate}}</span>
            <input type="text" formControlName="information">
        </label>
        <app-invalid-input 
        *ngIf="information.dirty && (information.errors?.information || information.status === 'VALID')" 
        message="{{ information.status === 'VALID' ? ('user-details.user-form.messages.success' | translate) 
        : ('user-details.user-form.messages.information' | translate) }}"
        type="{{ information.status === 'VALID' ? 'success' : 'error'}}"></app-invalid-input>
    </div>
    <div>
        <label>
            <span>{{'user-details.user-form.dateOfNotification' | translate}}</span>
            <input type="text" formControlName="dateOfNextNotif" placeholder="DD-MMM-YY">
        </label>
        <app-invalid-input 
        *ngIf="dateOfNextNotif.dirty && (dateOfNextNotif.errors?.date || dateOfNextNotif.status === 'VALID')" 
        message="{{ dateOfNextNotif.status === 'VALID' ? ('user-details.user-form.messages.success' | translate) 
        : ('user-details.user-form.messages.dateOfNextNotif' | translate) }}"
        type="{{ dateOfNextNotif.status === 'VALID' ? 'success' : 'error'}}"></app-invalid-input>
    </div>
    <input type="submit" [disabled]="userForm.pristine || userForm.status !== 'VALID'" value="{{'user-details.user-form.submit' | translate}}">
</form>
