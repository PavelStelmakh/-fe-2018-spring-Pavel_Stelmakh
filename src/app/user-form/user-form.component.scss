:host {    
    text-align: center; 
}

label {
    width: 100%;
    display: block;   
}

form > div {
    margin-bottom: 60px;
}

span {
    float: left;
    margin-left: 30px;    
    color: #ccc;
    font-family: Verdana;
}

.loading {
    position: absolute;
    width: 450px;
    margin-left: 25px;
    margin-top: 5px;
    font-family: Verdana;
}