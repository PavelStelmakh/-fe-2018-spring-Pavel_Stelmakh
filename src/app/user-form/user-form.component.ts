import { Component, OnInit, Output, EventEmitter } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { usernameValidator, userageValidator, dateValidator } from '../validators';
import { messageList } from '../shared/messageList';
import { IUserForm } from '../shared/IUserForm';

@Component({
  selector: 'app-user-form',
  templateUrl: './user-form.component.html',
  styleUrls: ['./user-form.component.scss']
})
export class UserFormComponent implements OnInit {
  @Output() submitForm: EventEmitter<IUserForm> = new EventEmitter<IUserForm>();
  userForm: FormGroup;
  messageList: object;

  constructor(private builder: FormBuilder) { }

  ngOnInit() {
    this.messageList = messageList;
    this.userForm = this.builder.group({
      username: ['', Validators.required, usernameValidator()],
      userage: ['', [Validators.required, userageValidator()]],
      birthday: ['', [Validators.required, dateValidator('YYYY/MM/DD')]],
      dateOfLogin: ['', [Validators.required, dateValidator('DD MMMM YYYY')]],
      DateOfNotification: ['', [Validators.required, dateValidator('DD-MMM-YY')]]
    });
  }

  get username() {
    return this.userForm.get('username');
  }

  get userage() {
    return this.userForm.get('userage');
  }

  get birthday() {
    return this.userForm.get('birthday');
  }

  get DateOfNotification() {
    return this.userForm.get('DateOfNotification');
  }

  get dateOfLogin() {
    return this.userForm.get('dateOfLogin');
  }

  onSubmit() {
    this.submitForm.emit(this.userForm.value);
  }

}
