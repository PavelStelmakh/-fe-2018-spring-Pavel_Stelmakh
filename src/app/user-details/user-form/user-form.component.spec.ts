import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { RouterTestingModule } from "@angular/router/testing";
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { UserFormComponent } from './user-form.component';
import { BASE_URL } from '../../url.service';
import { Observable, of } from 'rxjs';
import { TranslateModule, TranslateLoader, TranslateService } from '@ngx-translate/core';
import { Store } from '@ngrx/store';

const translations = {
  "user-details": {
    "user-form": {
      "name": "Your name",
      "password": "Your password",
      "age": "Your age",
      "birthday": "Your birthday",
      "dateOfLogin": "Date of login",
      "information": "Information",
      "dateOfNotification": "Date of notification",
      "submit": "send request",
      "messages": {
          "success": "OK",
          "name": "Username must consist of one or two word with camel case sensitive",
          "password": "Password have to contain min 3 char",
          "existname": "This name exist already",
          "age": "Range of numbers should from 18 till 65",
          "dateOfBirth": "date should have format YYYY/MM/DD",
          "dateOfFirstLogin": "date should have format DD MMMM YYYY",
          "information": "Field is required",
          "dateOfNextNotif": "date should have format DD-MMM-YY"
      }
  }
  }
};

class FakeLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<{}> {
    return of(translations);
  }
}

describe('UserFormComponent', () => {
  let fixture: ComponentFixture<UserFormComponent>;
  let translate: TranslateService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        HttpClientModule,
        HttpClientTestingModule,
        ReactiveFormsModule,
        TranslateModule.forRoot({
          loader: {provide: TranslateLoader, useClass: FakeLoader}
        })
      ],
      declarations: [
        UserFormComponent
      ],
      providers: [
        {provide: BASE_URL, useValue: 'http://localhost:3000'},
        {
          provide: Store,
          useValue: {
            dispatch: jest.fn(),
            pipe: jest.fn(() => of({}))
          }
        }
      ],
      schemas: [
       NO_ERRORS_SCHEMA
      ]
    }).compileComponents();

  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserFormComponent);
    translate = TestBed.get(TranslateService);
  });

  it('renders markup to snapshot', () => {
    translate.use('en');
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

});
