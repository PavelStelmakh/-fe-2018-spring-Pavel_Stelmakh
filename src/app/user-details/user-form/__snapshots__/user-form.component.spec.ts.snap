// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserFormComponent renders markup to snapshot 1`] = `
<app-user-form
  builder={[Function FormBuilder]}
  id="undefined"
  isCreateNewUser="false"
  popup={[Function PopupService]}
  storeProfile={[Function Object]}
  storeUsers={[Function Object]}
  update={[Function EventEmitter]}
  user={[Function UsersService]}
  userForm={[Function FormGroup]}
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <div>
      <label>
        <span>
          Your name
        </span>
        <input
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="name"
          ng-reflect-name="name"
          placeholder="Anna Maria"
          type="text"
        />
      </label>
      
      
      
    </div>
    
    <div>
      <label>
        <span>
          Your age
        </span>
        <input
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="age"
          ng-reflect-name="age"
          placeholder="18-65"
          type="text"
        />
      </label>
      
    </div>
    <div>
      <label>
        <span>
          Your birthday
        </span>
        <input
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="dateOfBirth"
          ng-reflect-name="dateOfBirth"
          placeholder="YYYY/MM/DD"
          type="text"
        />
      </label>
      
    </div>
    <div>
      <label>
        <span>
          Date of login
        </span>
        <input
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="dateOfFirstLogin"
          ng-reflect-name="dateOfFirstLogin"
          placeholder="DD MMMM YYYY"
          type="text"
        />
      </label>
      
    </div>
    <div>
      <label>
        <span>
          Information
        </span>
        <input
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="information"
          ng-reflect-name="information"
          type="text"
        />
      </label>
      
    </div>
    <div>
      <label>
        <span>
          Date of notification
        </span>
        <input
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="dateOfNextNotif"
          ng-reflect-name="dateOfNextNotif"
          placeholder="DD-MMM-YY"
          type="text"
        />
      </label>
      
    </div>
    <input
      disabled=""
      type="submit"
      value="send request"
    />
  </form>
</app-user-form>
`;
